sound/soc/sof/intel/hda.c

```diff
@@ -1117,14 +1119,144 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
 
 #if IS_ENABLED(CONFIG_SND_SOC_SOF_INTEL_SOUNDWIRE)
 
+static bool is_endpoint_present(struct sdw_slave *sdw_device,
+				struct asoc_sdw_codec_info *dai_info, int dai_type)
+{
+	int i;
+
+	for (i = 0; i < sdw_device->sdca_data.num_functions; i++) {
+		if (dai_type == dai_info->dais[i].dai_type)
+			return true;
+	}
+	dev_dbg(&sdw_device->dev, "Endpoint DAI type %d not found\n", dai_type);
+	return false;
+}
+
+static struct snd_soc_acpi_adr_device *find_acpi_adr_device(struct device *dev,
+							    struct sdw_slave *sdw_device,
+							    struct snd_soc_acpi_link_adr *link,
+							    int *amp_index)
+{
+	struct snd_soc_acpi_adr_device *adr_dev;
+	int index = link->num_adr;
+	int ep_index = 0;
+	int i, j;
+
+	link->mask = BIT(sdw_device->bus->link_id);
+	/* index is 0 based, we need allocate index + 1 for the array size */
+	if (!index)
+		adr_dev = devm_kzalloc(dev, sizeof(*adr_dev), GFP_KERNEL);
+	else
+		adr_dev = devm_krealloc(dev, (struct snd_soc_acpi_adr_device *)link->adr_d,
+					(index + 1) * sizeof(*adr_dev), GFP_KERNEL);
+
+	if (!adr_dev)
+		return NULL;
+
+	for (i = 0; i < asoc_sdw_get_codec_info_list_count(); i++) {
+		if (sdw_device->id.part_id != codec_info_list[i].part_id)
+			continue;
+
+		int amp_group_id = 1;
+
+		adr_dev[index].endpoints =
+			devm_kzalloc(dev, codec_info_list[i].dai_num *
+				     sizeof(struct snd_soc_acpi_endpoint), GFP_KERNEL);
```

Comment:
Is the problem here? If instead of assigning the endpoint pointer in adr_dev directly what if you made a local non-const pointer, then once you have filled in the endpoints you can then assign that to the const pointer in the adr_dev structure.

Reply:
No, the problem is
```c
struct snd_soc_acpi_adr_device {
        const u64 adr;
        const u8 num_endpoints;
        const struct snd_soc_acpi_endpoint *endpoints;
        const char *name_prefix;
};
```
and
```c
struct snd_soc_acpi_link_adr {
        const u32 mask;
        const u32 num_adr;
        const struct snd_soc_acpi_adr_device *adr_d;
};
```
I will get `assignment of read-only member` error even if the adr_device and the link_adr themselves are not const.